## Reading in data available through Oryzabase (Integrated Rice Science Database)
The purpose of this notebook is to read in and do a preliminary analysis of the data related to text descriptions and ontology term annotations that are available through Oryzabase. The data needs to be organized and also restructured into a standard format that will allow it to be easily combined with datasets from other resources. This notebook takes the following input files that were obtained from OryzaBase and produces a set of files that have standard columns, including the species name, gene names, gene synonyms, text descriptions, and ontology term annotations. The gene names column includes unique gene accessions, names, symbols, or identifiers. The gene synonyms column is included for strings that are not necessary unique identifiers for a particular gene but still refer to that gene or describe its function.

### Files read
```
plant-data/databases/oryzabase/OryzabaseGeneListEn_20190826010113.txt
```

### Files created
```
plant-data/reshaped_data/oryzabase_phenotype_descriptions_and_annotations.csv
```

In [1]:
import matplotlib.pyplot as plt
import matplotlib as mpl
import seaborn as sns
import sys
import os
import warnings
import pandas as pd
import numpy as np
import itertools
import re
import matplotlib.pyplot as plt
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize

sys.path.append("../../oats")
from oats.utils.utils import to_abbreviation
from oats.utils.constants import NCBI_TAG
from oats.utils.constants import EVIDENCE_CODES
from oats.nlp.preprocess import add_prefix
from oats.nlp.preprocess import other_delim_to_bar_delim
from oats.nlp.preprocess import get_ontology_ids
from oats.nlp.preprocess import concatenate_descriptions
from oats.nlp.preprocess import remove_punctuation
from oats.nlp.preprocess import handle_synonym_in_parentheses
from oats.nlp.preprocess import remove_enclosing_brackets
from oats.nlp.preprocess import remove_short_tokens
from oats.nlp.preprocess import concatenate_with_bar_delim

OUTPUT_DIR = "../reshaped_data"
mpl.rcParams["figure.dpi"] = 200
warnings.simplefilter('ignore')
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

### Part 1: Phenotypic Text Data
There are several columns that contain information about gene names and accessions. We need to know what type of information is in each in order to know which should be retained or parsed to form the desired cleaned dataset. We are interested in both gene names that should map to a specific accession (like cms-54257) as well as gene names that are enzyme descriptions (like Ubiquitin-Specific Protease) that could map to more than one gene in a particular species. Each type of information is valuable, but needs to be differentiated so that when comparing whether two rows are specifying the same gene, this is not confused with specifying two different genes that have the same function.

The gene symbols in this dataset are typically surrounded by square brackets but not always. If a second symbol for that same gene is mentioned in the same column, the second symbol might be enclosed in parentheses. The synonynms for the gene symbols are similarly sometimes enclosed in square brackets, and are typically separated by commas in cases where more than one are mentioned in this column. Also note that an underscore is being used to represent missing data, so this has to handled so that that character is not treated as a gene name that appears many times.

In [2]:
filename = "../databases/oryzabase/OryzabaseGeneListEn_20190826010113.txt"
usecols = ["CGSNL Gene Symbol", "Gene symbol synonym(s)", "CGSNL Gene Name", "Gene name synonym(s)", "Protein Name", 
           "Allele", "Explanation", "Trait Class", "RAP ID", "MUS ID", "Gramene ID", "Gene Ontology", 
           "Trait Ontology", "Plant Ontology"]
df = pd.read_table(filename, usecols=usecols, sep="\t")
df.fillna("", inplace=True)
unique_values = {col:len(pd.unique(df[col].values)) for col in df.columns}
print(df.shape)
for k,v in unique_values.items():
    print("{:25}{:8}".format(k,v))

(17674, 14)
CGSNL Gene Symbol           10041
Gene symbol synonym(s)      15948
CGSNL Gene Name              9845
Gene name synonym(s)        15216
Protein Name                 5588
Allele                       1151
Explanation                 13841
Trait Class                   726
RAP ID                      11531
MUS ID                      11642
Gramene ID                   1656
Gene Ontology                7624
Trait Ontology               2476
Plant Ontology                797


In [3]:
print(df[["CGSNL Gene Symbol", "Gene symbol synonym(s)"]].head(5))
print(df[["CGSNL Gene Symbol", "Gene symbol synonym(s)"]].sample(5))

    CGSNL Gene Symbol                  Gene symbol synonym(s)
0         [CMS-54257]               [cms-54257]*, [cms-54257]
1  [CMS-AK]([CMS-JP])  [cms-ak]([cms-jp]), [cms-jp], [cms-ak]
2           [CMS-ARC]   [cms-ARC]*, [cms-ARC] [mt], [cms-ARC]
3            [CMS-BO]                                [cms-bo]
4            [CMS-CW]                                [cms-CW]
      CGSNL Gene Symbol Gene symbol synonym(s)
13947                 _         Osthi11, thi11
8548                  _                 atpE.p
11023                 _               R2R3-MYB
3770               PPT2                 OsPPT2
3404              NAAT1                OsNAAT1


The values in the gene name synonym(s) column can be comma delimited lists if more than one synonym for the gene name was known. Sometimes quotes are used. Empty strings and possibly underscores can be used to denote missing information.

In [4]:
print(df[["CGSNL Gene Name"]].head(5))
print(df[["CGSNL Gene Name"]].sample(5))

                    CGSNL Gene Name
0  CYTOPLASMIC MALE STERILITY 54257
1     CYTOPLASMIC MALE STERILITY AK
2    CYTOPLASMIC MALE STERILITY ARC
3     CYTOPLASMIC MALE STERILITY BO
4     CYTOPLASMIC MALE STERILITY CW
                        CGSNL Gene Name
8795                                  _
4447   RING DOMAIN-CONTAINING PROTEIN 5
16059           GERMIN-LIKE PROTEIN 9-2
12753                                 _
15843                                 _


The gene name column has strings representing the full name of each gene rather than just the symbol. Note that an underscore is also being used to denote missing values in this column as well.

In [5]:
print(df[["Gene name synonym(s)"]].head(5))
print(df[["Gene name synonym(s)"]].sample(5))

                                Gene name synonym(s)
0  Cytoplasmic mutant induced by somaclonal varia...
1            Akebono' cytoplasm, 'Akebono' cytoplasm
2     ARC13829-16 cytoplasm, `ARC13829-26' cytoplasm
3  Chinsurah boro II' cytoplasm, `Chinsurah boro ...
4                        Chinese wild rice cytoplasm
                                    Gene name synonym(s)
15863  MATH-BTB protein 39, MDC protein with a BTB do...
13286         hybrid proline- or glycine-rich protein 10
14106                                                   
3263                                     spotted leaf 21
2347                          dynamin-related protein 3A


The values in the gene name synonym(s) column can be comma delimited lists if more than one synonym for the gene name was known. Sometimes quotes are used. Empty strings and possibly underscores can be used to denote missing information.

In [6]:
print(df[["Protein Name","Allele"]].sample(5))

                    Protein Name Allele
1479                                   
1007                                   
8675                                   
1194                                   
5239  TRYPTOPHAN DECARBOXYLASE 2       


Both the protein name and allele columns are sparse within the dataset. Either can be a single value or a comma delimited list of values. These may not need to be retained for finding reference to genes in other resources because we already have more standardized representations of the genes in other columns.

In [7]:
print(df[["RAP ID","MUS ID"]].sample(5))

             RAP ID            MUS ID
10335                                
4976   Os05g0103200  LOC_Os05g01270.1
8239   Os11g0181500  LOC_Os11g07916.1
1539                                 
3100                                 


Both the RAP ID and the MUS ID can columns can contain multiple values for a given gene which are included as members of a comma delimited list. These values can also be missing using the same scheme for missing information as in the rest of the dataset.

The following functions were created based on the needs following how the gene symbols, names, synonyms, and accessions are previously described in this dataset. These are a not guaranteed to be a perfectly accurate method of parsing in the information in this dataset but they are meant to approximate what is required based on going through the dataset by hand. The methods are meant to be applied only to specific columns within the dataset, and to make the code that later cleans the columns more readable by compressing multiple cleaning steps into a single function. Some of these rely on other very specific functions that are within the text preprocessing module and not shown here.

In [8]:
def clean_oryzabase_symbol(string):
    # Should be applied to the gene symbol column in the dataset.
    # Returns a single string representing a bar delimited list of gene symbols.
    string = string.replace("*","")
    names = handle_synonym_in_parentheses(string, min_length=4)
    names = [remove_enclosing_brackets(name) for name in names]
    names = remove_short_tokens(names, min_length=2)
    names_string = concatenate_with_bar_delim(*names)
    return(names_string)

def clean_oryzabase_symbol_synonyms(string):
    # Should be applied to the gene symbol synonym(s) column in the dataset.
    # Returns a single string representing a bar delimited list of gene symbols.
    string = string.replace("*","")
    names = string.split(",")
    names = [name.strip() for name in names]
    names = [remove_enclosing_brackets(name) for name in names]
    names_string = concatenate_with_bar_delim(*names)
    return(names_string)

### Part 2: Ontology Term Annotations
Multiple columns within the dataset specify ontology term annotations that have been applied to the geen mentioned on that particular line. Ontology term annotations are separated into different columns based on which ontology the terms belong to, and both the term ID of each annotation and the accompanying label for that term and explicitly given. Columns for terms from the Gene Ontolgoy (GO), Plant Ontology (PO), and Plant Trait Ontology (TO) are all included. There is no information about what evidence codes these annotations are associated with in this dataset.

In [9]:
print(df[["Gene Ontology"]].head(5))
print(df[["Plant Ontology"]].head(5))
print(df[["Trait Ontology"]].head(5))

                                       Gene Ontology
0  GO:0000001 - mitochondrion inheritance, GO:000...
1  GO:0000001 - mitochondrion inheritance, GO:000...
2  GO:0007275 - multicellular organismal development
3  GO:0007275 - multicellular organismal developm...
4  GO:0007275 - multicellular organismal development
                                      Plant Ontology
0                               PO:0009066 - anther 
1                               PO:0009066 - anther 
2  PO:0009082 - spikelet floret , PO:0020048 - mi...
3  PO:0009082 - spikelet floret , PO:0020048 - mi...
4  PO:0009082 - spikelet floret , PO:0020048 - mi...
                                      Trait Ontology
0  TO:0000232 - cytoplasmic male sterility (sensu...
1  TO:0000232 - cytoplasmic male sterility (sensu...
2  TO:0000232 - cytoplasmic male sterility (sensu...
3  TO:0000232 - cytoplasmic male sterility (sensu...
4  TO:0000232 - cytoplasmic male sterility (sensu...


Both the term IDs and labels for each annotation are given. Multiple annotations from the same ontology for a given line are separated by commas. We want to parse out just the gene ontology IDs for the cleaned dataset so that they can be referenced later, all the other information is not needed. There is a function to the return a list of gene IDs present in a longer string of text that is in the preprocessing module.

### Part 3: Handling other text description and keyword information in this data
This dataset does not contain any columns that consistly contain a natural language description of a phenotype associated with a given gene. But some text-based information is still present. The trait class column contains a value from a limited set of keyword descriptors for the trait a particular gene is associated with. The size of the vocabulary used is obtained here. Also see the specific description of this keyword vocabulary here (https://shigen.nig.ac.jp/rice/oryzabase/traitclass/). The explaination column also occasionally contains text information about a corresponding phenotype.

In [10]:
# Get a list sorted by number of occurences for each trait class.
description_counts = df["Trait Class"].value_counts().to_dict()
sorted_tuples = sorted(description_counts.items(), key = lambda x: x[1], reverse=True)
for t in sorted_tuples[0:10]:
    print("{:6}  {:20}".format(t[1],t[0][:70]))

  4121                      
  3621   Biochemical character
  2566   Other              
   875   Tolerance and resistance - Stress tolerance
   862   Biochemical character,  Tolerance and resistance - Stress tolerance
   527   Tolerance and resistance - Disease resistance
   293   Reproductive organ - Spikelet, flower, glume, awn
   244   Biochemical character,  Tolerance and resistance - Disease resistance
   222   Character as QTL - Yield and productivity
   204   Tolerance and resistance - Stress tolerance,  Other


The most common value in the trait class column is whitespace or an empty string indicating missing data. Another very common value though is 'Other' which has 2,566 occurences out of the 17,674 total instances. This needs to be handled if using this information as text descriptions because this contains no semantics relevant to the phenotype (two phenotypes with trait classes of 'Other' should not be considered similar).

In [11]:
print(df[["Explanation"]].sample(30))

                                             Explanation
12203                     F-box protein. LOC_Os02g19410.
11498                                                   
5095                                                    
12025       LOC_Os10g05240. Os_F0350 in Hua et al. 2011.
10124                                                   
8168                                                    
13626                                     LOC_Os09g18230
3557                           LOC_Os03g29970. AB288040.
3263                                                    
16784  LOC_Os07g47300. Q0D3S3. TO:0006060: leaf chlor...
12477  D29710 (corresponds to Os09g0482100, Os09g0482...
7061                                                    
6034        LOC_Os03g60650. Q84T94. AJ575237. EU368725. 
7038                                                    
9898   13 kDa prolamin. Os05g0329200, LOC_Os05g26480 ...
5900                                     LOC_Os04g55660.
4465   Efflux transporter of ph

The explanation column holds information potentially about the phenotype, but also sometimes contains redundant information about the gene names or identifiers and sometimes the ontology term annotations as well. Sometimes methods are mentioned as well. Some of this could be handled with parsing to remove the redundant information that already appears somewhere else in a particular column for this line, but this should be considered irregular text annotations or descriptions for the purposes of downstream analyses. The following cell contains a preliminary attempt at a function that cleans values in this column by removing some redundant information from other columns.

In [12]:
def clean_oryzabase_explainations(string):
    # Should be applied to the explanation column in the dataset.
    # Returns a version of the the value in that column without some redundant information.
    ontology_ids = get_ontology_ids(string)
    for ontology_id in ontology_ids:
        string = string.replace(ontology_id,"")
        string = remove_punctuation(string)
    return(string)

In [13]:
# Restructuring and combining columns that have gene name information.
df["CGSNL Gene Symbol"] = df["CGSNL Gene Symbol"].apply(clean_oryzabase_symbol)
df["Gene symbol synonym(s)"] = df["Gene symbol synonym(s)"].apply(clean_oryzabase_symbol_synonyms)
df["CGSNL Gene Name"] = df["CGSNL Gene Name"].apply(lambda x: x.replace("_","").strip())
df["Gene name synonym(s)"] = df["Gene name synonym(s)"].apply(lambda x: other_delim_to_bar_delim(string=x, delim=","))
df["gene_names"] = np.vectorize(concatenate_with_bar_delim)(df["RAP ID"], df["MUS ID"], df["CGSNL Gene Symbol"], df["Gene symbol synonym(s)"], df["CGSNL Gene Name"], df["Gene name synonym(s)"])

# Restructuring and combining columns that have ontology term annotations.
df["Gene Ontology"] = df["Gene Ontology"].apply(lambda x: concatenate_with_bar_delim(*get_ontology_ids(x))) 
df["Trait Ontology"] = df["Trait Ontology"].apply(lambda x: concatenate_with_bar_delim(*get_ontology_ids(x))) 
df["Plant Ontology"] = df["Plant Ontology"].apply(lambda x: concatenate_with_bar_delim(*get_ontology_ids(x))) 
df["term_ids"] = np.vectorize(concatenate_with_bar_delim)(df["Gene Ontology"], df["Trait Ontology"], df["Plant Ontology"])

# Adding other expected columns and subsetting the dataset.
df["species"] = "osa"
df["description"] = df["Explanation"].apply(clean_oryzabase_explainations)
df["pmid"] = ""
df = df[["species", "gene_names", "description", "term_ids"]]
print(df[["species","gene_names"]].head(10))
print(df[["species","gene_names"]].sample(10))

  species                                         gene_names
0     osa  CMS-54257|cms-54257|CYTOPLASMIC MALE STERILITY...
1     osa  CMS-JP|CMS-AK|[cms-ak]([cms-jp])|cms-jp|cms-ak...
2     osa  CMS-ARC|cms-ARC|cms-ARC] [mt|CYTOPLASMIC MALE ...
3     osa  CMS-BO|cms-bo|CYTOPLASMIC MALE STERILITY BO|Ch...
4     osa  CMS-CW|cms-CW|CYTOPLASMIC MALE STERILITY CW|Ch...
5     osa  CMS-GAM|cms-GAM|CYTOPLASMIC MALE STERILITY GAM...
6     osa  CMS-HL|cms-HL|CYTOPLASMIC MALE STERILITY HL|HL...
7     osa  CMS-IR66707A|cms-IR66707A|CYTOPLASMIC MALE STE...
8     osa  CMS-KALINGA-I|cms-Kalinga-I|CYTOPLASMIC MALE S...
9     osa  CMS-KHIABORO|cms-Khiaboro|CYTOPLASMIC MALE STE...
      species                                         gene_names
3691      osa  Os12g0278800|LOC_Os12g18120.4, LOC_Os12g18120....
5863      osa  Os01g0650900|LOC_Os01g46210.1|GELP22|OsGELP22|...
16262     osa                      Os07g0118300|LOC_Os07g02710.1
6552      osa  Os01g0637800|LOC_Os01g45090.1|OsMYB8|myb trans...
1363

In [14]:
# Outputting the dataset of descriptions to a csv file.
path = os.path.join(OUTPUT_DIR,"oryzabase_phenotype_descriptions_and_annotations.csv")
df.to_csv(path, index=False)